lucene/core/src/java/org/apache/lucene/index/VectorValues.java

```diff
@@ -74,6 +74,18 @@ public BytesRef binaryValue() throws IOException {
     throw new UnsupportedOperationException();
   }
 
+  /**
+   * Return the k nearest neighbor documents as determined by comparison of their vector values
+   * for this field, to the given vector, by the field's search strategy. If the search strategy is
+   * reversed, lower values indicate nearer vectors, otherwise higher scores indicate nearer
+   * vectors. Unlike relevance scores, vector scores may be negative.
+   * @param target the vector-valued query
+   * @param k      the number of docs to return
+   * @param fanout control the accuracy/speed tradeoff - larger values give better recall at higher cost
```

Comment:
Yeah, I think there needs to be a follow-on exposing the index-time controls, which indeed are much more potent than this search-time fanout, which has only a small impact on recall and latency. In this patch they are globals in HnswGraphBuilder, but there is no API for setting them. I am thinking the index-time hyperparameters would be specified in IndexWriterConfig?

Reply:
> Yeah, I think there needs to be a follow-on exposing the index-time controls, which indeed are much more potent than this search-time fanout, which has only a small impact on recall and latency. In this patch they are globals in HnswGraphBuilder, but there is no API for setting them.

OK, makes sense.

> I am thinking the index-time hyperparameters would be specified in IndexWriterConfig?

Hmm, maybe these could be codec level controls?  Or maybe `FieldInfo`?  They would be per-vector-field configuration right?

Reply:
Yeah that's a good point. While experimenting with GloVe I'm learning that different settings are appropriate for different vectors, so field-level control might be needed. I'm not sure how codec-level controls are exposed. Don't Codecs get created automatically using no-args constructors and service autodiscovery? Did you mean something like perFieldVectorFormat? Except I doubt we need a new format; it's more about some metadata values that we would store in the field, so I think yeah it would go in FieldInfo. But I'm reluctant to expose hnsw-specific hyperparameters in `VectorField`, which we want to support other algorithms as well. Maybe this is a good use case for `IndexableField.getAttributes()`?

Reply:
Yeah this probably needs to be field level. Different strokes for different collections of vectors. I'm not sure how to expose since the parameters will be different for different ANN implementations. Might be a good use case for generic IndexedField.attributes?

Reply:
> Don't Codecs get created automatically using no-args constructors and service autodiscovery? 

They do at read (search) time!  But at write time, you can pass parameters that alter how the Codec does its work, as long as the resulting index is then readable at search time with no-args constructors.

I vaguely remember talking about having ways for Codec at read-time to also take options, but I'm not sure that was ever fully designed / pushed ... @s1monw may remember?

> But I'm reluctant to expose hnsw-specific hyperparameters in VectorField, which we want to support other algorithms as well.
> Might be a good use case for generic IndexedField.attributes?

Yeah, maybe?  I agree it is not obvious where the API should live and how it then finds its way into the ANN data structure construction when writing each segment.

Reply:
@mikemccand it was pushed but removed again in https://issues.apache.org/jira/browse/LUCENE-9257 a little while ago. I get why it's removed but it seems useful maybe we add it back?

Reply:
Ahh OK thanks for the context @s1monw.